Let's Encrypt SSL certificate support

mejo- opened the pull request:
This PR adds new tasks to obtain and install Let's Encrypt SSL certificates for pveproxy using the external role [systemli.letsencrypt](https://galaxy.ansible.com/systemli/letsencrypt/).

lae commented on tasks/ssl_letsencrypt.yml:
```diff
@@ -0,0 +1,19 @@
+---
+
```
can you remove this newline?

lae commented on tasks/ssl_letsencrypt.yml:
```diff
@@ -0,0 +1,19 @@
+---
+
+- name: Install Proxmox Let's Encrypt post-hook script
```
I don't think it's necessary to create and use a post-hook from the included role here. This can easily be taken care of with a `copy` task and calling the appropriate handler (like in `ssl_config.yml`).

mejo- replied:
Unfortunately not. The Let's Encrypt certificate will be renewed automatically every three months. Afterwards, the post-hook script needs to be executed (certbot will take care of this).

lae replied:
Ah, right, I forgot.

mejo- commented:
I can confirm that I tested this on my setup. I use the role `systemli.letsencrypt` a lot.

mejo- commented:
Here's a log:

```
PLAY [proxmox] *****************************************************************************************************************************

TASK [proxmox : Install Proxmox Let's Encrypt post-hook script] ****************************************************************************
--- before
+++ after: /tmp/tmpLemND3/pve-letsencrypt-post-hook.sh.j2
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+cp -a /etc/letsencrypt/live/pve1.example.org/privkey.pem   /etc/pve/local/pveproxy-ssl.key
+cp -a /etc/letsencrypt/live/pve1.example.org/fullchain.pem /etc/pve/local/pveproxy-ssl.pem
+service pveproxy restart

changed: [pve1]
--- before
+++ after: /tmp/tmpPPMdtu/pve-letsencrypt-post-hook.sh.j2
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+cp -a /etc/letsencrypt/live/pve2.example.org/privkey.pem   /etc/pve/local/pveproxy-ssl.key
+cp -a /etc/letsencrypt/live/pve2.example.org/fullchain.pem /etc/pve/local/pveproxy-ssl.pem
+service pveproxy restart

changed: [pve2]

TASK [proxmox : Request Let's Encrypt certificate for pve2.example.org] **********************************************************************

TASK [jnv.debian-backports : add distribution-specific variables] **************************************************************************

TASK [jnv.debian-backports : add backports repository] *************************************************************************************

TASK [letsencrypt : install Let's Encrypt Certbot client] **********************************************************************************

TASK [letsencrypt : install certbot plugin 'apache' on webservers] *************************************************************************

TASK [letsencrypt : install certbot DNS challenge helper script] ***************************************************************************

TASK [letsencrypt : create directory /etc/letsencrypt/keys] ********************************************************************************

TASK [letsencrypt : install certbot DNS challenge nsupdate key] ****************************************************************************

TASK [letsencrypt : install certbot DNS challenge nsupdate private key] ********************************************************************

TASK [letsencrypt : add system group 'letsencrypt'] ****************************************************************************************

TASK [letsencrypt : adjust permissions for certificate directories] ************************************************************************

TASK [letsencrypt : check if letsencrypt_account_email is set] *****************************************************************************

TASK [letsencrypt : check if a Let's Encrypt account exists] *******************************************************************************

TASK [letsencrypt : create Let's Encrypt account] ******************************************************************************************

TASK [letsencrypt : prepare authenticator options for apache] ******************************************************************************

TASK [letsencrypt : prepare authenticator options for standalone] **************************************************************************
ok: [pve2]
ok: [pve1]

TASK [letsencrypt : prepare authenticator options for webroot] *****************************************************************************

TASK [letsencrypt : prepare cert name option] **********************************************************************************************
ok: [pve2]
ok: [pve1]

TASK [letsencrypt : prepare optional test cert option] *************************************************************************************
ok: [pve2]
ok: [pve1]

TASK [letsencrypt : prepare post-hook options] *********************************************************************************************
ok: [pve2]
ok: [pve1]

TASK [letsencrypt : prepare post-hook options] *********************************************************************************************

TASK [letsencrypt : register Let's Encrypt certificate with HTTP challenge] ****************************************************************
changed: [pve1]
changed: [pve2]

TASK [letsencrypt : register Let's Encrypt certificate with DNS challenge] *****************************************************************

TASK [letsencrypt : read in system users] **************************************************************************************************
changed: [pve2]
changed: [pve1]

TASK [letsencrypt : add users to letsencrypt system group] *********************************************************************************
```

mejo- commented:
I tried to address all the the remaining issues you raised.

Unfortunately I don't see an easy way to test this function either with the testsuite you use. There's no way do define role dependencies, is there? Probably you would have to use a more sophisticated testsuite, like molecule, for that.

mejo- commented:
About testing: sorry, forget my last comment. The real problem will be to verify the hostname when requesting a Let's Encrypt certificate. Unfortunately even the staging/testing environment of Let's Encrypt requires hostname validation.

lae commented:
Yeah, hostname validation is the problem I was alluding to as not being simple to test. As for adding dependencies, they're specified in `.travis.yml`. (I have my own thoughts about Molecule.)

mejo- commented:
Wow, that was fast! Thanks a ton :)
